lfs/client.go

```diff
@@ -125,6 +126,58 @@ func (b *byteCloser) Close() error {
 	return nil
 }
 
+type upload struct {
+	OIDPath  string
+	Filename string
+	CB       CopyCallback
+}
+
+type UploadQueue struct {
+	uc     chan upload
+	ec     chan *WrappedError
+	wg     sync.WaitGroup
+	errors []*WrappedError
+}
+
+func NewUploadQueue(workers int) *UploadQueue {
+	q := &UploadQueue{uc: make(chan upload, workers), ec: make(chan *WrappedError)}
+
+	go func() {
+		for err := range q.ec {
+			q.errors = append(q.errors, err)
+		}
+	}()
+
+	for i := 0; i < workers; i++ {
+		go func(n int) {
+			for upload := range q.uc {
+				q.wg.Add(1)
+				err := Upload(upload.OIDPath, upload.Filename, upload.CB)
+				if err != nil {
+					q.ec <- err
+				}
+				q.wg.Done()
+			}
+		}(i)
+	}
+
+	return q
+}
+
+func (q *UploadQueue) Upload(oidPath, filename string, cb CopyCallback) {
+	q.uc <- upload{oidPath, filename, cb}
+}
+
+func (q *UploadQueue) Wait() {
+	close(q.uc)
+	q.wg.Wait()
+	close(q.ec)
```

Comment:
Just to clarify, closing these channels will stop the for loops in the goroutines once they've processed the last messages.  Right?  Not that it matters much since the command will exit right after this happens.

Also, I think `Wait()` is a strange name.  It makes sense here, looking at the implementation of the queue.  But I don't want the upload queue in the push command to _wait_, I want it to _process_ or _run_ or something.  Prior art: http://golang.org/pkg/os/exec/#Cmd.Run